### PR TITLE
Generate WASM/JS Fallback during build process

### LINKED ISSUE #6768: Add JS fallback for WASM components
Some of our current crypto algorithms are compiled to WASM (argon2, liboqs). This causes issues in environments where WASM is disabled (browsers on iOS with Lockdown Mode, Tor-browser and derivatives in some security settings).

In the future we would like to compile Rust-based SDK to WASM so we will rely on WASM even more.

We would like to allow those special browser cases to still use Tuta.

## Tasks
- Argon2
   - [x] Try wasm-to-JS transpiler manually for one of our WASM libraries
   - [x] Integrate with build system to produce the fallback
   - [x] Implement WASM detection at runtime
   - [x] Load fallback when needed
 - Liboqs
   - [x] Try wasm-to-JS transpiler manually for one of our WASM libraries
   - [x] Integrate with build system to produce the fallback
   - [x] Implement WASM detection at runtime
   - [x] Load fallback when needed

## Test Notes
See #6835

### LINKED ISSUE #6835: Use esbuild to generate wasm fallback

### PR DESCRIPTION
**Pull request by murilopereirame**
Generates WASM artifact and it's fallback during the build without the need of check in the WASM files.

Closes #6835

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -49,6 +51,28 @@ jobs:
       - name: lint, formatting
         run: |
           npm run check
+      - name: get emscripten cached location and version
+        run: |
+          echo "emscripten_path=$(pwd)/emsdk" >> $GITHUB_ENV
+          echo "emscripten_version=$(curl https://api.github.com/repos/emscripten-core/emsdk/tags | jq -r '.[0].name')" >> $GITHUB_ENV
```
Comment (charlag) [inline .github/workflows/test.yml:57]: I would just hardcode the version and maybe put it in the build documentation/requirements

File: package.json [hunk 2]
```diff
@@ -45,7 +45,7 @@
 		"@types/qrcode-svg": "1.1.1",
 		"@types/systemjs": "6.1.1",
 		"@types/winreg": "1.2.31",
-		"better-sqlite3": "git+https://github.com/tutao/better-sqlite3-sqlcipher#c6ee076a4f15671d8a0fa7ddc16d12b179deb350",
+		"better-sqlite3": "git+https://github.com/tutao/better-sqlite3-sqlcipher#7e6d8b36cd804b4d5d07264c9a2f4a8820881586",
```
Comment (charlag) [inline package.json:48]: why did this one change?

File: packages/tuta-wasm-loader/lib/WasmHandler.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import { exec } from "node:child_process"
+import * as util from "node:util"
+
+export interface WasmGeneratorOptions {
+	workingDir?: string
+	env?: Record<string, any>
+}
+
+export interface FallbackGeneratorOptions extends WasmGeneratorOptions {
+	optimizationLevel: string
```
Comment (charlag) [inline packages/tuta-wasm-loader/lib/WasmHandler.ts:11]: optimization is only used for wasm2js, right? Maybe worth documenting it

File: packages/tuta-wasm-loader/lib/index.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import * as fs from "node:fs"
+
+export interface Library {
+	name: string
+	command: string
+	options: FallbackGeneratorOptions
+}
+
+export interface LoadOptions {
+	// Tool needed to transpile the wasm file to a JavaScript file
```
Comment (charlag) [inline packages/tuta-wasm-loader/lib/index.ts:13]: use jsdoc comments

File: packages/tuta-wasm-loader/lib/index.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+				return await generateWasmFallback(wasmPath, {
+					optimizationLevel: lib.options.optimizationLevel,
+					env: lib.options.env,
+					wasm2jsPath: lib.options.wasm2jsPath,
+				})
+			}
+		},
+		async transform(source: string, id: string) {
+			if (id.includes(".wasm")) {
```
Comment (charlag) [inline packages/tuta-wasm-loader/lib/index.ts:130]: why is this part needed? sorry, I probably asked already before but better to add a comment

File: packages/tutanota-crypto/test/KyberTest.ts [hunk 6]
```diff
@@ -1,12 +1,39 @@
...
+import { decapsulate, encapsulate, generateKeyPair, LibOQSExports } from "../lib/encryption/Liboqs/Kyber.js"
 import { random } from "../lib/index.js"
-
-const liboqs = await loadWasmModuleFromFile("../lib/encryption/Liboqs/liboqs.wasm")
+import { loadWasmModuleFallback, loadWasmModuleFromFile } from "./WebAssemblyTestUtils.js"
+import { $ } from "zx"
+import fs from "node:fs"
 
 o.spec("Kyber", function () {
+	// We need to generate the webassembly stuff during tests
```
Comment (charlag) [inline packages/tutanota-crypto/test/KyberTest.ts:9]: Even with comment I don't get it. Don't we also invoke it from the test build script?

File: types/globals.d.ts [hunk 7]
```diff
@@ -20,3 +20,9 @@ declare type IdTuple = Readonly<[Id, Id]>
 declare type Writeable<T> = { -readonly [P in keyof T]: T[P] }
 
 declare type None = null | undefined
+
+declare module "*.wasm" {
+	const loadWasm: (options?: { forceFallback?: boolean }) => Promise<Argon2IDExports | LibOQSExports>
```
Comment (charlag) [inline types/globals.d.ts:25]: I think the narrowing to two exports doesn't help because we need to cast the result anyway so I would stay with generic wasm exports

File: package.json (hunk 2 above)
Comment (murilopereirame) [reply package.json:48]: Rebase https://github.com/tutao/tutanota/pull/6842

File: packages/tutanota-crypto/test/KyberTest.ts (hunk 6 above)
Comment (murilopereirame) [reply packages/tutanota-crypto/test/KyberTest.ts:9]: Tests under packages run before app's test, so no esbuild yet and consequently, no generated wasm or fallback

File: packages/tuta-wasm-loader/lib/index.ts (hunk 5 above)
Comment (murilopereirame) [reply packages/tuta-wasm-loader/lib/index.ts:130]: Returning the generated loader or fallback. It's called after load and should return the JS code if the Id is already generated/imported or call **load** again to handle imports

https://rollupjs.org/plugin-development/#build-hooks

Comment (charlag) [reply packages/tuta-wasm-loader/lib/index.ts:130]: Please add this comment to the code

Comment (murilopereirame) [reply packages/tuta-wasm-loader/lib/index.ts:137]: Done

File: packages/tutanota-crypto/test/KyberTest.ts (hunk 6 above)
Comment (murilopereirame) [reply packages/tutanota-crypto/test/KyberTest.ts:9]: Done

File: packages/tuta-wasm-loader/lib/index.ts (hunk 5 above)
Comment (charlag) [reply packages/tuta-wasm-loader/lib/index.ts:137]: Okay damn I didn't read it properly, I didn't want to know what `transform()` does, I can look it up, I wanted to know why we need that condition for wasm files. Is it so that rollup doesn't try to transform them?

Comment (murilopereirame) [reply packages/tuta-wasm-loader/lib/index.ts:137]: > Okay damn I didn't read it properly, I didn't want to know what `transform()` does, I can look it up, I wanted to know why we need that condition for wasm files. Is it so that rollup doesn't try to transform them?

I check if it includes wasm in the name to prevent our plugin from processing non-wasm files. All project files goes through the plugin lifecycle (resolveId, load and transform) and I just ignore the ones that isn't related to WebAssembly stuff.

Comment (charlag) [reply packages/tuta-wasm-loader/lib/index.ts:137]: but you always return `null` so you never process anything, why do we need to add this callback at all? You can just remove `transform` if i get it right? https://rollupjs.org/plugin-development/#a-simple-example

Comment (murilopereirame) [reply packages/tuta-wasm-loader/lib/index.ts:137]: So, during the rollup I replace the wasm call by **wasm-loader:\<lib\>.wasm** and I want to return the source code for the loader inside the transform, that's the reason for doing this if check, returning the source code **ONLY** for loader (**wasm-loader:\<lib\>.wasm**) and fallback (**wasm-fallback:\<lib\>.wasm**) and not for all the other files. Sorry for the confusion

**Review by charlag — APPROVED** [review]
Looks good now, thanks! let's squash it before merging tho

**Comment by murilopereirame** [comment]
> Looks good now, thanks! let's squash it before merging tho

Squashed and rebased
